crates/bevy_animation/src/lib.rs

```diff
@@ -397,194 +493,222 @@ struct AnimationTargetContext<'a> {
-        self.animation.seek_time
+    /// Returns the [`ActiveAnimation`] associated with the given animation
+    /// node if it's currently playing.
+    ///
+    /// If the animation isn't currently active, returns `None`.
+    pub fn animation(&self, animation: AnimationNodeIndex) -> Option<&ActiveAnimation> {
+        self.playing_animations.get(&animation)
     }
 
-    /// Seek to a specific time in the animation.
-    pub fn seek_to(&mut self, seek_time: f32) -> &mut Self {
-        self.animation.seek_time = seek_time;
-        self
+    /// Returns a mutable reference to the [`ActiveAnimation`] associated with
+    /// the given animation node if it's currently active.
+    ///
+    /// If the animation isn't currently active, returns `None`.
+    pub fn animation_mut(&mut self, animation: AnimationNodeIndex) -> Option<&mut ActiveAnimation> {
+        self.playing_animations.get_mut(&animation)
     }
 
-    /// Reset the animation to its initial state, as if no time has elapsed.
-    pub fn replay(&mut self) {
-        self.animation.replay();
+    /// Returns true if the animation is currently playing or paused, or false
+    /// if the animation is stopped.
+    pub fn animation_is_playing(&self, animation: AnimationNodeIndex) -> bool {
+        self.playing_animations.contains_key(&animation)
     }
 }
 
 /// A system that advances the time for all playing animations.
 pub fn advance_animations(
     time: Res<Time>,
     animation_clips: Res<Assets<AnimationClip>>,
-    mut players: Query<&mut AnimationPlayer>,
+    animation_graphs: Res<Assets<AnimationGraph>>,
+    mut players: Query<(&mut AnimationPlayer, &Handle<AnimationGraph>)>,
 ) {
-    for mut player in players.iter_mut() {
-        let paused = player.paused;
-        if paused {
+    let delta_seconds = time.delta_seconds();
+    for (mut player, graph_handle) in players.iter_mut() {
+        let Some(animation_graph) = animation_graphs.get(graph_handle) else {
             continue;
-        }
-
-        // Advance the main animation.
-        if let Some(animation_clip) = animation_clips.get(&player.animation.animation_clip) {
-            player
-                .animation
-                .update(time.delta_seconds(), animation_clip.duration);
         };
 
-        // Advance transition animations.
-        player.transitions.retain_mut(|transition| {
-            // Decrease weight. Expire the transition if necessary.
-            transition.current_weight -= transition.weight_decline_per_sec * time.delta_seconds();
-            if transition.current_weight <= 0.0 {
-                return false;
+        let AnimationPlayer {
+            ref mut playing_animations,
+            ref blend_weights,
+            ref mut schedule,
+            ..
+        } = *player;
+
+        // Tick animations, and schedule them.
+
+        schedule.clear();
+
+        let mut dfs = animation_graph.dfs();
+        let mut weights = HashMap::new();
```

Comment:
Would it be possible to pre-allocate hashmap memory here to avoid frequent allocation?

Reply:
Dont